api/web3server.go

```diff
@@ -119,54 +148,54 @@ func init() {
 }
 
 // NewWeb3Server creates a new web3 server
-func NewWeb3Server(core CoreService, httpPort int, cacheURL string, queryLimit uint64) *Web3Server {
-	svr := &Web3Server{
-		web3Server: &http.Server{
-			Addr: ":" + strconv.Itoa(httpPort),
-		},
+func NewWeb3Server(core CoreService, cacheURL string, opts ...Web3Option) *Web3Server {
+	web3 := &Web3Server{
 		coreService: core,
-		queryLimit:  queryLimit,
+		cache:       newAPICache(15*time.Minute, cacheURL),
 	}
-
-	mux := http.NewServeMux()
-	mux.Handle("/", svr)
-	svr.web3Server.Handler = mux
-	svr.cache = newAPICache(15*time.Minute, cacheURL)
-	return svr
+	op := &svrOption{}
+	for _, opt := range opts {
+		opt(op)
+	}
+	if op.httpPort != 0 {
+		web3.httpSvr = NewHTTPServer(op.httpPort, web3)
+	}
+	if op.websocketPort != 0 {
+		web3.websocketSvr = NewWebSocketServer(op.websocketPort, web3)
+	}
+	return web3
 }
 
 // Start starts the API server
 func (svr *Web3Server) Start(_ context.Context) error {
-	go func() {
-		if err := svr.web3Server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
-			log.L().Fatal("Node failed to serve.", zap.Error(err))
+	if svr.httpSvr != nil {
+		if err := svr.httpSvr.Start(context.Background()); err != nil {
+			return err
+		}
+	}
+	if svr.websocketSvr != nil {
+		if err := svr.websocketSvr.Start(context.Background()); err != nil {
+			return err
 		}
-	}()
+	}
 	return nil
 }
 
 // Stop stops the API server
 func (svr *Web3Server) Stop(_ context.Context) error {
-	return svr.web3Server.Shutdown(context.Background())
-}
-
-func (svr *Web3Server) ServeHTTP(w http.ResponseWriter, req *http.Request) {
-	switch req.Method {
-	case "POST":
-		httpResp := svr.handlePOSTReq(req)
-
-		// write results into http reponse
-		w.Header().Set("Content-Type", "application/json; charset=UTF-8")
-		if err := json.NewEncoder(w).Encode(httpResp); err != nil {
-			log.L().Warn("fail to respond request.")
-		}
-	default:
-		w.WriteHeader(http.StatusMethodNotAllowed)
+	var err error
+	if svr.httpSvr != nil {
+		err = svr.httpSvr.Stop(context.Background())
+	}
+	if svr.websocketSvr != nil {
+		err = svr.websocketSvr.Stop(context.Background())
```

Comment:
1. stop in reverse order.
2. if line 188 returns an error, it will be overwritten by line 191